cube/cube.c

```diff
@@ -4181,34 +4184,42 @@ static void demo_init_vk(struct demo *demo) {
     } else {
         /* Try to auto select most suitable device */
         if (demo->gpu_number == -1) {
-            uint32_t count_device_type[VK_PHYSICAL_DEVICE_TYPE_CPU + 1];
-            memset(count_device_type, 0, sizeof(count_device_type));
-
             VkPhysicalDeviceProperties physicalDeviceProperties;
+            int prev_priority = 0;
             for (uint32_t i = 0; i < gpu_count; i++) {
                 vkGetPhysicalDeviceProperties(physical_devices[i], &physicalDeviceProperties);
                 assert(physicalDeviceProperties.deviceType <= VK_PHYSICAL_DEVICE_TYPE_CPU);
-                count_device_type[physicalDeviceProperties.deviceType]++;
-            }
 
-            VkPhysicalDeviceType search_for_device_type = VK_PHYSICAL_DEVICE_TYPE_DISCRETE_GPU;
-            if (count_device_type[VK_PHYSICAL_DEVICE_TYPE_DISCRETE_GPU]) {
-                search_for_device_type = VK_PHYSICAL_DEVICE_TYPE_DISCRETE_GPU;
-            } else if (count_device_type[VK_PHYSICAL_DEVICE_TYPE_INTEGRATED_GPU]) {
-                search_for_device_type = VK_PHYSICAL_DEVICE_TYPE_INTEGRATED_GPU;
-            } else if (count_device_type[VK_PHYSICAL_DEVICE_TYPE_VIRTUAL_GPU]) {
-                search_for_device_type = VK_PHYSICAL_DEVICE_TYPE_VIRTUAL_GPU;
-            } else if (count_device_type[VK_PHYSICAL_DEVICE_TYPE_CPU]) {
-                search_for_device_type = VK_PHYSICAL_DEVICE_TYPE_CPU;
-            } else if (count_device_type[VK_PHYSICAL_DEVICE_TYPE_OTHER]) {
-                search_for_device_type = VK_PHYSICAL_DEVICE_TYPE_OTHER;
-            }
+                // Continue next gpu if this gpu does not support the surface.
+                VkBool32 supported = 0;
+                vkGetPhysicalDeviceSurfaceSupportKHR(physical_devices[i], 0, demo->surface, &supported);
+                if (!supported) continue;
 
-            for (uint32_t i = 0; i < gpu_count; i++) {
-                vkGetPhysicalDeviceProperties(physical_devices[i], &physicalDeviceProperties);
-                if (physicalDeviceProperties.deviceType == search_for_device_type) {
```

Comment:
Use priority to select physical device so that only one loop is needed. It will be more convenient to filter GPU if only one loop existed.